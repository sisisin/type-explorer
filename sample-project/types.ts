type Baz = number;
type Bar = {
  a: string[] | string;
  baz: Baz;
};

type Bar1 = { b: string[] };

export type Foo = {
  x: string;
  bar?: Bar;
};
