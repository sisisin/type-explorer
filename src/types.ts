export type TreeNode = {
  id: number;
  variableName: string | undefined;
  typeName: string;
  children?: TreeNode[];
};

export type TreeNodeLike = {
  variableName: string | undefined;
  typeName: string;
  children?: TreeNodeLike[];
};
