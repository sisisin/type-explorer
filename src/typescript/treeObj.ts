import { TreeNode } from '../types';

export const treeObj: TreeNode = {
  id: 1,
  variableName: 'Foo',
  typeName: undefined,
  children: [
    {
      id: 2,
      variableName: 'x',
      typeName: 'string',
    },
    {
      id: 3,
      variableName: 'bar',
      typeName: 'Bar',
      children: [
        {
          id: 4,
          variableName: 'a',
          typeName: 'string[] | string',
          children: [
            { id: 5, variableName: 'string[]', typeName: 'string[]' },
            { id: 6, variableName: 'string', typeName: 'string' },
          ],
        },
        {
          id: 7,
          variableName: 'baz',
          typeName: 'Baz',
          children: [{ id: 8, variableName: 'number', typeName: undefined }],
        },
      ],
    },
  ],
};
