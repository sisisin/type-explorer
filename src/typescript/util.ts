import * as ts from 'typescript';

function getKindName(node: ts.Node) {
  return ts.SyntaxKind[node?.kind];
}
export function getDescendantAtPos(node: ts.Node, pos: number) {
  let ret: ts.Node | undefined;

  while (true) {
    const next: ts.Node | undefined = getChildAtPos(ret || node, pos);
    if (next == null) return ret;
    ret = next;
  }
}
export function contain(node: ts.Node, pos: number) {
  return node.pos <= pos && pos < node.end;
}
export function getChildAtPos(node: ts.Node, pos: number) {
  if (contain(node, pos)) {
    for (const child of node.getChildren()) {
      if (contain(child, pos)) return child;
    }
  }
  return undefined;
}
