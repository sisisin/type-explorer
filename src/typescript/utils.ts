import * as ts from 'typescript';

function getKindName(node: ts.Node) {
  return ts.SyntaxKind[node?.kind];
}
export function contain(node: ts.Node, pos: number) {
  return node.pos <= pos && pos < node.end;
}
export function getDescendantAtPos(node: ts.Node, pos: number) {
  let ret: ts.Node | undefined;

  while (true) {
    const next: ts.Node | undefined = getChildAtPos(ret || node, pos);
    if (next == null) return ret;
    ret = next;
  }
}
export function getChildAtPos(node: ts.Node, pos: number) {
  if (contain(node, pos)) {
    for (const child of node.getChildren()) {
      if (contain(child, pos)) return child;
    }
  }
  return undefined;
}
export function forEachChildArray(node: ts.Node): ts.Node[] {
  const nodes: ts.Node[] = [];
  node.forEachChild((n) => nodes.push(n));
  return nodes;
}
export function isPrimitiveKeyword(node: ts.Node) {
  switch (node.kind) {
    case ts.SyntaxKind.BooleanKeyword:
    case ts.SyntaxKind.StringKeyword:
    case ts.SyntaxKind.NumberKeyword:
    case ts.SyntaxKind.SymbolKeyword:
    case ts.SyntaxKind.NullKeyword:
    case ts.SyntaxKind.UndefinedKeyword:
    case ts.SyntaxKind.AnyKeyword:
    case ts.SyntaxKind.UnknownKeyword:
    case ts.SyntaxKind.BigIntKeyword:
    case ts.SyntaxKind.ObjectKeyword:
    case ts.SyntaxKind.ThisKeyword:
    case ts.SyntaxKind.VoidKeyword:
    case ts.SyntaxKind.NeverKeyword:
    case ts.SyntaxKind.LiteralType:
      return true;

    default:
      return false;
  }
}

function isTypeDefinitionNode(node: ts.Node) {
  return (
    isPrimitiveKeyword(node) ||
    ts.isTypeLiteralNode(node) ||
    ts.isTypeReferenceNode(node) ||
    ts.isUnionTypeNode(node) ||
    ts.isArrayTypeNode(node) ||
    ts.isPropertySignature(node)
  );
}

export function isSyntaxList(node: ts.Node): node is ts.SyntaxList {
  return node.kind === ts.SyntaxKind.SyntaxList;
}

/**
 * Find type definition node among children nodes
 * @example
 * type Foo = { foo: Bar }  // => returned `{ foo: Bar }`
 * @param node A type definition node
 */
export function findDeclarationNode(node: ts.Node): ts.Node | undefined {
  return node.getChildren().find(isTypeDefinitionNode);
}

export function findDeclarationNodes(node: ts.Node): ts.Node[] {
  return node.getChildren().filter(isTypeDefinitionNode);
}
