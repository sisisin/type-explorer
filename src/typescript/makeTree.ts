import * as ts from 'typescript';
import { TreeNode } from '../types';
import { assert, getIdGenerator, isNonNullable } from '../utils';
import {
  findDeclarationNode,
  findDeclarationNodes,
  getDescendantAtPos,
  isPrimitiveKeyword,
  isSyntaxList,
} from './utils';

export function makeTree(program: ts.Program, src: ts.SourceFile, pos: number) {
  const checker = program.getTypeChecker();
  const node = getDescendantAtPos(src, pos);
  if (node === undefined) {
    console.log(`cannot find Node that pos is ${pos}`);
    return;
  }
  const treeNodeStartingPoint = findTreeNodeStartingPoint(node);
  if (treeNodeStartingPoint === undefined) {
    console.log(`Invalid Node: '${node.getText()}' kind: '${ts.SyntaxKind[node.kind]}'`);
    return;
  }

  return from(
    { checker, genId: getIdGenerator(), root: treeNodeStartingPoint },
    treeNodeStartingPoint,
  );

  function isTreeNodeStartingPoint(node: ts.Node) {
    return (
      ts.isPropertySignature(node) ||
      ts.isTypeAliasDeclaration(node) ||
      ts.isInterfaceDeclaration(node) ||
      ts.isVariableDeclaration(node) // note: it is unsupported `VariableDeclarationList`
    );
  }
  function findTreeNodeStartingPoint(node: ts.Node): ts.Node | undefined {
    if (ts.isSourceFile(node)) return undefined;
    else if (isTreeNodeStartingPoint(node)) return node;

    return findTreeNodeStartingPoint(node.parent);
  }
}

export interface Context {
  genId: () => number;
  checker: ts.TypeChecker;
  root: ts.Node;
}

function from(ctx: Context, node: ts.Node): TreeNode | undefined {
  const { genId, checker, root } = ctx;

  if (ts.isTypeAliasDeclaration(node)) {
    const declarationBody = findDeclarationNode(node);

    return {
      id: genId(),
      typeName: node.name.getText(),
      variableName: undefined,
      ...makeChildren(ctx, declarationBody),
    };
  } else if (isPrimitiveKeyword(node)) {
    return {
      id: genId(),
      typeName: node.getText(),
      variableName: undefined,
    };
  } else if (ts.isPropertySignature(node)) {
    const declarationBody = findDeclarationNode(node);
    return {
      id: genId(),
      typeName: node.type?.getText() ?? assert('TypeNode must not be undefined'),
      variableName: node.name.getText(),
      ...makeChildren(ctx, declarationBody),
    };
  } else if (ts.isTypeReferenceNode(node)) {
    return fromTypeReferenceNode(ctx, node);
  } else if (ts.isArrayTypeNode(node)) {
    const declarationBody = findDeclarationNode(node);
    return {
      id: genId(),
      typeName: node.getText(),
      variableName: undefined,
      ...makeChildren(ctx, declarationBody),
    };
  } else if (ts.isInterfaceDeclaration(node)) {
    const syntaxList = node.getChildren().find(isSyntaxList);
    if (syntaxList === undefined) throw new Error(`${node.name.getText()} has not SyntaxList`);
    const children = findDeclarationNodes(syntaxList)
      .map((body) => from(ctx, body))
      .filter(isNonNullable);
    const childrenObj: Pick<TreeNode, 'children'> = {
      children: children.length > 0 ? children : undefined,
    };
    return {
      id: genId(),
      typeName: node.name.getText(),
      variableName: undefined,
      ...childrenObj,
    };
  } else if (ts.isVariableDeclaration(node)) {
    return fromVariableDeclaration(ctx, node);
  } else if (ts.isTypeLiteralNode(node)) {
    return {
      id: genId(),
      typeName: 'Anonymous(Object Literal)',
      variableName: (root as ts.VariableDeclaration).name.getText(),
      ...makeChildren(ctx, node),
    };
  } else if (ts.isPropertyAssignment(node)) {
    return {
      id: genId(),
      typeName: checker.typeToString(checker.getTypeAtLocation(node)),
      variableName: node.name.getText(),
    };
  }

  return undefined;
}

function fromVariableDeclaration(ctx: Context, node: ts.VariableDeclaration) {
  const { genId, checker } = ctx;
  const type = checker.getTypeAtLocation(node);

  // case of `const foo: Foo = ...`
  if (node.type) {
    const reference = type.aliasSymbol?.getDeclarations()?.[0]!;
    const decl = findDeclarationNode(reference);
    return {
      id: genId(),
      typeName: node.type.getText(),
      variableName: node.name.getText(),
      ...makeChildren(ctx, decl),
    };
  }

  const typeNode = checker.typeToTypeNode(type);
  if (typeNode === undefined) return undefined;

  // case of `let x = 1; // => inferred a number type`
  if (isPrimitiveKeyword(typeNode)) {
    return {
      id: genId(),
      typeName: checker.typeToString(type),
      variableName: node.name.getText(),
    };
  }

  // case of `const obj = {...}`. Right side expression is TypeLiteral node.
  if (ts.isTypeLiteralNode(typeNode)) {
    const obj = node.forEachChild((n) => (ts.isObjectLiteralExpression(n) ? n : undefined));
    if (obj === undefined)
      throw new Error(
        `Unexpected undefined. TypeLiteralNode should have a ObjectLiteralExpression`,
      );

    return {
      id: genId(),
      typeName: 'Anonymous(Object Literal)',
      variableName: node.name.getText(),
      ...makeChildren(ctx, obj),
    };
  }
}
function fromTypeReferenceNode(ctx: Context, node: ts.TypeReferenceNode) {
  // todo: improve get node
  const { checker } = ctx;
  const identifier = node.getChildren()[0];
  const symbol = checker.getSymbolAtLocation(identifier);
  const declNode = symbol?.getDeclarations()?.[0];
  if (declNode === undefined) return undefined;
  return from(ctx, declNode);
}

function makeChildren(ctx: Context, node: ts.Node | undefined): Pick<TreeNode, 'children'> {
  const children = makeChildrenList(ctx, node);
  return children.length > 0 ? { children } : {};
}

function makeChildrenList(ctx: Context, node: ts.Node | undefined): TreeNode[] {
  if (node === undefined) return [];
  if (isPrimitiveKeyword(node)) {
    return [from(ctx, node)].filter(isNonNullable);
  } else if (ts.isTypeLiteralNode(node)) {
    return node.members.map((m) => from(ctx, m)).filter(isNonNullable);
  } else if (ts.isTypeReferenceNode(node)) {
    return [from(ctx, node)].filter(isNonNullable);
  } else if (ts.isUnionTypeNode(node)) {
    const syntaxList = node.getChildren()[0];
    if (isSyntaxList(syntaxList)) {
      return findDeclarationNodes(syntaxList)
        .map((body) => from(ctx, body))
        .filter(isNonNullable);
    }
  } else if (ts.isArrayTypeNode(node)) {
    return [from(ctx, node)].filter(isNonNullable);
  } else if (isSyntaxList(node)) {
    return findDeclarationNodes(node)
      .map((body) => from(ctx, body))
      .filter(isNonNullable);
  } else if (ts.isTypeLiteralNode(node)) {
    return node.members.map((m) => from(ctx, m)).filter(isNonNullable);
  } else if (ts.isObjectLiteralExpression(node)) {
    return node.properties.map((p) => from(ctx, p)).filter(isNonNullable);
  }

  return [];
}
