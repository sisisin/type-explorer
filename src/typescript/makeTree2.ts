import * as ts from 'typescript';
import { TreeNode } from '../types';
import { getDescendantAtPos } from './util';
import { isNonNullable } from '../utils';

function getIdGenerator() {
  let id = 0;
  return () => ++id;
}

export function makeTree(program: ts.Program, src: ts.SourceFile, pos: number) {
  const checker = program.getTypeChecker();
  const node = getDescendantAtPos(src, pos);
  if (node === undefined) {
    console.log(`cannot find Node that pos is ${pos}`);
    return;
  }
  const treeNodeStartingPoint = findTreeNodeStartingPoint(node);
  if (treeNodeStartingPoint === undefined) {
    console.log(`Invalid Node: '${node.getText()}' kind: '${ts.SyntaxKind[node.kind]}'`);
    return;
  }

  return makeTypeTree({ checker, genId: getIdGenerator() }, treeNodeStartingPoint);
}

interface Context {
  genId: () => number;
  checker: ts.TypeChecker;
}
function makeTypeTree(ctx: Context, node: ts.Node): TreeNode | undefined {
  const { checker, genId } = ctx;

  if (ts.isTypeAliasDeclaration(node)) {
    const base = {
      id: genId(),
      typeName: node.name.getText(), // type Some = ... <- `Some` can access `node.name`
      variableName: undefined,
    };
    const declarationBody = findDeclarationBody(node);
    if (declarationBody) {
      return {
        ...base,
        children: [makeTypeTree(ctx, declarationBody)].filter(isNonNullable),
      };
    } else {
      return base;
    }
  } else if (isPrimitiveKeyword(node)) {
    return {
      id: genId(),
      typeName: node.getText(),
      variableName: undefined,
    };
  }
}

function isPrimitiveKeyword(node: ts.Node) {
  switch (node.kind) {
    case ts.SyntaxKind.BooleanKeyword:
    case ts.SyntaxKind.StringKeyword:
    case ts.SyntaxKind.NumberKeyword:
    case ts.SyntaxKind.SymbolKeyword:
      return true;

    default:
      return false;
  }
}

function isTreeNodeStartingPoint(node: ts.Node): node is TreeNodeStartingPoint {
  return ts.isPropertySignature(node) || ts.isTypeAliasDeclaration(node);
}
type TreeNodeStartingPoint = ts.PropertySignature | ts.TypeAliasDeclaration;

function findTreeNodeStartingPoint(node: ts.Node): TreeNodeStartingPoint | undefined {
  if (ts.isSourceFile(node)) return undefined;
  else if (isTreeNodeStartingPoint(node)) return node;

  return findTreeNodeStartingPoint(node.parent);
}

function findDeclarationBody(node: ts.Node): ts.Node | undefined {
  return node.getChildren().find((child) => {
    return isPrimitiveKeyword(child);
  });
}
