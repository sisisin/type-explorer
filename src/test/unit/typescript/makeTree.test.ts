import * as ts from 'typescript';
import { TreeNodeLike } from '../../../types';
import { makeTree } from '../../../typescript/makeTree';
import { createProgram, dropId, getArgPart } from './helpers';

let p: ts.Program;

beforeAll(() => {
  p = createProgram('./fixtures');
});
describe('Unsupported Node', () => {
  it('should return undefined that node is unsupported', () => {
    const { f, pos } = getArgPart(p, 'unsupported.ts', 'Unsupported__Namespace');
    expect(makeTree(p, f, pos)).toBeUndefined();
  });
});

describe('primitive typed TypeAliasDeclaration', () => {
  function getPrimitiveTreeNode(targetType: string, childType: string) {
    return {
      typeName: targetType,
      variableName: undefined,
      children: [
        {
          typeName: childType,
          variableName: undefined,
        },
      ],
    };
  }
  const parameters = [
    { identifier: 'AliasOfBoolean', childType: 'boolean' },
    { identifier: 'AliasOfString', childType: 'string' },
    { identifier: 'AliasOfNumber', childType: 'number' },
    { identifier: 'AliasOfSymbol', childType: 'symbol' },
    { identifier: 'AliasOfNull', childType: 'null' },
    { identifier: 'AliasOfUndefined', childType: 'undefined' },
    { identifier: 'AliasOfAny', childType: 'any' },
    { identifier: 'AliasOfUnknown', childType: 'unknown' },
    { identifier: 'AliasOfBigInt', childType: 'bigint' },
    { identifier: 'AliasOfObject', childType: 'object' },
    { identifier: 'AliasOfVoid', childType: 'void' },
    { identifier: 'AliasOfNever', childType: 'never' },
    { identifier: 'AliasOfStringLiteral', childType: "'a_string_literal'" },
  ];

  parameters.forEach(({ identifier, childType }) => {
    it(`should make tree from ${childType}`, () => {
      const tree = getPrimitiveTreeNode(identifier, childType);
      const { f, pos } = getArgPart(p, 'aliased-primitives.ts', identifier);
      const actual = makeTree(p, f, pos);
      expect(dropId(actual)).toStrictEqual(tree);
    });
  });
  it('should make tree from exported type', () => {
    const tree = getPrimitiveTreeNode('ExportedType', 'unknown');
    const { f, pos } = getArgPart(p, 'exported-type.ts', 'ExportedType');
    const actual = makeTree(p, f, pos);
    expect(dropId(actual)).toStrictEqual(tree);
  });
});

describe('Literal Object typed TypeAliasDeclaration', () => {
  it(`should make tree from PropertySignature`, () => {
    const tree: TreeNodeLike = {
      typeName: 'string',
      variableName: 'foo',
      children: [
        {
          typeName: 'string',
          variableName: undefined,
        },
      ],
    };
    const { f, pos } = getArgPart(p, 'property-signature.ts', 'foo');
    const actual = makeTree(p, f, pos);
    expect(dropId(actual)).toStrictEqual(tree);
  });

  it('should make tree from TypeAlias', () => {
    const tree: TreeNodeLike = {
      typeName: 'FooObject',
      variableName: undefined,
      children: [
        {
          typeName: 'string',
          variableName: 'foo',
          children: [
            {
              typeName: 'string',
              variableName: undefined,
            },
          ],
        },
        {
          typeName: 'AliasOfSomething',
          variableName: 'bar',
          children: [
            {
              typeName: 'AliasOfSomething',
              variableName: undefined,
              children: [
                {
                  typeName: 'symbol',
                  variableName: undefined,
                },
              ],
            },
          ],
        },
      ],
    };
    const { f, pos } = getArgPart(p, 'property-signature.ts', 'FooObject');
    const actual = makeTree(p, f, pos);
    expect(dropId(actual)).toStrictEqual(tree);
  });
});

describe('Union typed TypeAliasDeclaration', () => {
  it('should make tree which has Union Attributes', () => {
    const tree: TreeNodeLike = {
      typeName: 'U',
      variableName: undefined,
      children: [
        {
          typeName: 'Alias',
          variableName: undefined,
          children: [
            {
              typeName: 'string',
              variableName: undefined,
            },
          ],
        },
        {
          typeName: 'number',
          variableName: undefined,
        },
      ],
    };
    const { f, pos } = getArgPart(p, 'union.ts', 'U');
    const actual = makeTree(p, f, pos);
    expect(dropId(actual)).toStrictEqual(tree);
  });
});
describe('Array typed TypeAliasDeclaration', () => {
  it('should make tree which has Array type', () => {
    const tree: TreeNodeLike = {
      typeName: 'AliasOfArray',
      variableName: undefined,
      children: [
        {
          typeName: 'AliasOfBoolean[]',
          variableName: undefined,
          children: [
            {
              typeName: 'AliasOfBoolean',
              variableName: undefined,
              children: [{ typeName: 'boolean', variableName: undefined }],
            },
          ],
        },
      ],
    };
    const { f, pos } = getArgPart(p, 'aliased-array.ts', 'AliasOfArray');
    const actual = makeTree(p, f, pos);
    expect(dropId(actual)).toStrictEqual(tree);
  });
});

describe('InterfaceDeclaration', () => {
  const tree: TreeNodeLike = {
    typeName: 'FooInterface',
    variableName: undefined,
    children: [
      {
        typeName: 'string',
        variableName: 'foo',
        children: [
          {
            typeName: 'string',
            variableName: undefined,
          },
        ],
      },
    ],
  };
  it('should make tree from basic interface', () => {
    const { f, pos } = getArgPart(p, 'interface.ts', 'FooInterface');
    const actual = makeTree(p, f, pos);
    expect(dropId(actual)).toStrictEqual(tree);
  });
});

describe('FunctionDeclaration', () => {
  it('should make tree with explicitly return type', () => {
    const tree: TreeNodeLike = {
      variableName: 'something',
      typeName: 'string',
    };
    const { f, pos } = getArgPart(p, 'function.ts', 'something');
    const actual = makeTree(p, f, pos);
    expect(dropId(actual)).toStrictEqual(tree);
  });
  it('should make tree with implicitly return type', () => {
    pending(
      "unsupported inferred return type because I can't understand to get return type from node",
    );
    const tree: TreeNodeLike = {
      variableName: 'somethingInferred',
      typeName: 'boolean',
    };
    const { f, pos } = getArgPart(p, 'function.ts', 'somethingInferred');
    const actual = makeTree(p, f, pos);
    expect(dropId(actual)).toStrictEqual(tree);
  });
});
