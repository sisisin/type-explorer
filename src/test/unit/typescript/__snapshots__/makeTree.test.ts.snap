// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should make tree from Alias to primitive Type 1`] = `
Object {
  "children": Array [],
  "id": 1,
  "typeName": "AliasOfBoolean",
  "variableName": undefined,
}
`;

exports[`should make tree from PropertySignature 1`] = `
Object {
  "children": Array [
    Object {
      "id": 1,
      "typeName": "number",
      "variableName": "a",
    },
    Object {
      "id": 2,
      "typeName": "string",
      "variableName": "b",
    },
    Object {
      "children": Array [
        Object {
          "id": 4,
          "typeName": "boolean",
          "variableName": undefined,
        },
      ],
      "id": 3,
      "typeName": "AliasOfBoolean",
      "variableName": "c",
    },
  ],
  "id": 5,
  "typeName": "BarObject",
  "variableName": "bar",
}
`;

exports[`should make tree from PropertySignature which references primitive Alias 1`] = `
Object {
  "children": Array [
    Object {
      "id": 2,
      "typeName": "boolean",
      "variableName": undefined,
    },
  ],
  "id": 1,
  "typeName": "AliasOfBoolean",
  "variableName": "c",
}
`;

exports[`should make tree from TypeAliasDeclaration 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "id": 1,
          "typeName": "number",
          "variableName": undefined,
        },
        Object {
          "id": 2,
          "typeName": "number[]",
          "variableName": undefined,
        },
      ],
      "id": 3,
      "typeName": "number | number[]",
      "variableName": "foo",
    },
    Object {
      "children": Array [
        Object {
          "id": 4,
          "typeName": "number",
          "variableName": "a",
        },
        Object {
          "id": 5,
          "typeName": "string",
          "variableName": "b",
        },
        Object {
          "children": Array [
            Object {
              "id": 7,
              "typeName": "boolean",
              "variableName": undefined,
            },
          ],
          "id": 6,
          "typeName": "AliasOfBoolean",
          "variableName": "c",
        },
      ],
      "id": 8,
      "typeName": "BarObject",
      "variableName": "bar",
    },
  ],
  "id": 9,
  "typeName": "FooObject",
  "variableName": undefined,
}
`;

exports[`should make tree from Union with TypeAlias 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "id": 2,
          "typeName": "string",
          "variableName": undefined,
        },
      ],
      "id": 1,
      "typeName": "AliasOfString",
      "variableName": undefined,
    },
    Object {
      "id": 3,
      "typeName": "string[]",
      "variableName": undefined,
    },
  ],
  "id": 4,
  "typeName": "AliasOfString | string[]",
  "variableName": "baz",
}
`;
